doc/src/manual/interfaces.md

```diff
@@ -739,3 +739,63 @@ yields another `SparseVecStyle`, that its combination with a 2-dimensional array
 yields a `SparseMatStyle`, and anything of higher dimensionality falls back to the dense arbitrary-dimensional framework.
 These rules allow broadcasting to keep the sparse representation for operations that result
 in one or two dimensional outputs, but produce an `Array` for any other dimensionality.
+
+## [IO Stream](@id man-interface-iostream)
+
+| Required methods                          | Brief description                                                                               |
+|:----------------------------------------- |:----------------------------------------------------------------------------------------------- |
+| `unsafe_read(io, ::Ptr{UInt8}, ::UInt)`   | Copy bytes from the `IO` to a pointer. |
+| `unsafe_write(io, ::Ptr{UInt8}, ::UInt)`  | Copy bytes from a pointer to the `IO`. |
+| `eof(io)`                                 | Whether the IO stream is at an end.    |
+
+| Optional methods                          | Brief description                                                                               |
+|:----------------------------------------- |:----------------------------------------------------------------------------------------------- |
+| `read(io, ::Type{UInt8})`                 | Read a byte from the stream.           |
+| `write(io, ::UInt8)`                      | Write a byte to the stream.            |
+| `close(io)`                               | Close the stream.                      |
+| `seek(io, ::Integer)`                     | Seek to a specific position.           |
+
+To implement an `IO` object it is required to define the low-level methods `unsafe_read`
+and `unsafe_write` which enable copying of data between the `IO` and a location given by a
+pointer.  Additionally, `eof` is required and should return `true` if reading from the
+`IO` is valid, else `false`.
+
+An `IO` that only defines `unsafe_read`, `unsafe_write` and `eof` can have most `isbits`
+types written or read to or from it, but the `read(io, ::Type{UInt8})` and `write(io,
+::UInt8)` methods must be defined to enable reading and writing individual bytes.
```

Comment:
I think the ordering of thoughts here is a bit is a bit confusing, but also I am not very familiar with the IO interface (and am grateful that you're documenting it!)

My expectation from reading this is that `write(io, x::MyType)` should ultimately be (and in `Base`, is) written in terms of `write(io, x::UInt8)`, so this wording might be clearer:
"An `isbits` type T (for which `isbits(T)`) typically defines `write(::IO, ::T)` in terms of "simpler" `write` methods, ultimately ending in `write(io, ::UInt8)` (and equivalently for `read`). The `read(io, ::Type{UInt8})` and `write(io,
::UInt8)` methods must be defined to enable reading and writing individual bytes of these `isbits` types.

But perhaps more importantly, this doesn't match the exploration I just did:

To probe the system:

```julia
julia> struct MyIO <: IO
       end

julia> write(MyIO(), 1 + 2*im)
ERROR: MyIO does not support byte I/O
Stacktrace:
  [1] error(::Type, ::String)
    @ Base ./error.jl:42
  [2] write(s::MyIO, x::UInt8)
    @ Base ./io.jl:224
  [3] unsafe_write
    @ ./io.jl:238 [inlined]
  [4] unsafe_write
    @ ./io.jl:646 [inlined]
  [5] unsafe_write(s::MyIO, p::Base.RefValue{Int64}, n::Int64)
    @ Base ./io.jl:644
  [6] write
    @ ./io.jl:647 [inlined]
  [7] write
    @ ./io.jl:650 [inlined]
  [8] write
    @ ./io.jl:637 [inlined]
  [9] write(s::MyIO, z::Complex{Int64})
    @ Base ./complex.jl:221
 [10] top-level scope
    @ REPL[11]:1
```

The stacktrace is a bit uninformative, but here is the sequence that I can tell.

- [9] the complex number is written as its components
- [8] multiple arguments are written
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L638-L644
- [7] each component is written
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L651-L653
- [6] e.g. Int is written by making a `Ref{Int}` and writing that
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L649
- [5] now we make our first call to `unsafe_write`:
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L646-L647
- [4] which calls this:
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L648
- [3] which calls this:
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L235-L241
And this is the one of the required methods for the IO interface according to the PR at hand (https://github.com/JuliaLang/julia/pull/41291/files#diff-f7164005c91058bbacdf095d8de0f02dd6194daf17d58979114eb505d292e35bR748)

- [2] as just seen, there is a fallback for `unsafe_write` written in terms of `write(s::IO, x::UInt8)`, so ultimately this call sequence calls an erroring "fallback" (tangentially: instead of throwing a `ERROR: MethodError: no method matching `write(::MyIO, ::UInt8)` and using e.g. `Experimental.show_error_hints` to deliver a better error message.)
https://github.com/JuliaLang/julia/blob/e469a1e4da7e7356257107d174b63fb5206bee04/base/io.jl#L224

So in this case it would be sufficient to just define `Base.write(::MyIO, ::UInt8)`, and I'm not sure how to reconcile that with this line in the PR.

I thought I might get a bit more understanding by looking to see some of those definitions of `Base.write`, but I'm not sure what I was looking for. In any case, here they are:

```julia
julia> _fieldtype(x, i) = try
       fieldtypes(x)[i]
       catch
       missing
       end
_fieldtype (generic function with 1 method)

julia> filter(m -> _fieldtype(m.sig, 3) === UInt8, methods(write).ms)
[1] write(io::Union{Core.CoreSTDERR, Core.CoreSTDOUT}, x::UInt8) in Base at coreio.jl:26
[2] write(to::Base.GenericIOBuffer, a::UInt8) in Base at iobuffer.jl:445
[3] write(io::Base.AbstractPipe, byte::UInt8) in Base at io.jl:360
[4] write(pipe::Base64.Base64EncodePipe, x::UInt8) in Base64 at /Applications/Julia-1.6.app/Contents/Resources/julia/share/julia/stdlib/v1.6/Base64/src/encode.jl:101
[5] write(io::Base.SecretBuffer, b::UInt8) in Base at secretbuffer.jl:112
[6] write(s::Base.BufferStream, b::UInt8) in Base at stream.jl:1351
[7] write(s::Base.LibuvStream, b::UInt8) in Base at stream.jl:1091
[8] write(s::IOStream, b::UInt8) in Base at iostream.jl:366
[9] write(::Base.DevNull, ::UInt8) in Base at coreio.jl:16
[10] write(f::Base.Filesystem.File, c::UInt8) in Base.Filesystem at filesystem.jl:136
[11] write(s::IO, x::UInt8) in Base at io.jl:224
```

Reply:
To be honest, I find your wording much more confusing than mine, maybe somebody else wants to come along and suggest a 3rd alternative?  (I'm not attached to my original wording, I'm just unsure what to change.)

As far as defining only `write(io, ::UInt8)` goes... yes, I don't find that particularly surprising.  Like I had implied, I find the current situation rather confusing.  I think writing some types if only `write(io, ::UInt8)` is defined will crash?  Regardless, I think the only way to resolve this is to get help from someone more intimately familiar with `IO`.  So far as I can tell, what I have written in this PR cannot break.